In [1]:
import pandas as pd
import numpy as np

In [2]:
location1 = ('mcpupdpdpstcy2017.csv')
location2 = ('mcpupdpdpstcy2016.csv')

In [3]:
df17 = pd.read_csv(location1)
df16 = pd.read_csv(location2)

In [4]:
df17.shape

(1162898, 84)

In [5]:
df16.shape

(1131550, 84)

In [6]:
df17.columns

Index(['npi', 'nppes_provider_last_org_name', 'nppes_provider_first_name',
       'nppes_provider_mi', 'nppes_credentials', 'nppes_provider_gender',
       'nppes_entity_code', 'nppes_provider_street1', 'nppes_provider_street2',
       'nppes_provider_city', 'nppes_provider_zip5', 'nppes_provider_zip4',
       'nppes_provider_state', 'nppes_provider_country',
       'specialty_description', 'description_flag',
       'medicare_prvdr_enroll_status', 'total_claim_count',
       'total_30_day_fill_count', 'total_drug_cost', 'total_day_supply',
       'bene_count', 'ge65_suppress_flag', 'total_claim_count_ge65',
       'total_30_day_fill_count_ge65', 'total_drug_cost_ge65',
       'total_day_supply_ge65', 'bene_count_ge65_suppress_flag',
       'bene_count_ge65', 'brand_suppress_flag', 'brand_claim_count',
       'brand_drug_cost', 'generic_suppress_flag', 'generic_claim_count',
       'generic_drug_cost', 'other_suppress_flag', 'other_claim_count',
       'other_drug_cost', 'mapd_suppress

In [7]:
df16.columns

Index(['npi', 'nppes_provider_last_org_name', 'nppes_provider_first_name',
       'nppes_provider_mi', 'nppes_credentials', 'nppes_provider_gender',
       'nppes_entity_code', 'nppes_provider_street1', 'nppes_provider_street2',
       'nppes_provider_city', 'nppes_provider_zip5', 'nppes_provider_zip4',
       'nppes_provider_state', 'nppes_provider_country',
       'specialty_description', 'description_flag',
       'medicare_prvdr_enroll_status', 'total_claim_count',
       'total_30_day_fill_count', 'total_drug_cost', 'total_day_supply',
       'bene_count', 'ge65_suppress_flag', 'total_claim_count_ge65',
       'total_30_day_fill_count_ge65', 'total_drug_cost_ge65',
       'total_day_supply_ge65', 'bene_count_ge65_suppress_flag',
       'bene_count_ge65', 'brand_suppress_flag', 'brand_claim_count',
       'brand_drug_cost', 'generic_suppress_flag', 'generic_claim_count',
       'generic_drug_cost', 'other_suppress_flag', 'other_claim_count',
       'other_drug_cost', 'mapd_suppress

### Problem 1: Average number of beneficiaries per provider in 2017 (only include those with more than 10 beneficiaries)

In [8]:
# Method 1 [Problem 1]: direct mean function

df17['bene_count'].mean()

158.3494585173676

In [9]:
# Method 2 [Problem 1]: using groupby function

#Step 1 - calculate the average (mean) beneficiaries for each provider
benef_per_prov_df17_a = df17.groupby(df17['npi'])['bene_count'].mean()

#Step 2 - calculate the average (mean) beneficiaries of all providers
avg_benef_per_prov_a = benef_per_prov_df17_a.mean()

avg_benef_per_prov_a

158.3494585173676

In [10]:
# Method 3 [Problem 1]: pivot table function

# step 1 - calculate the average (mean) beneficiaries for each provider
benef_per_prov_df17_b = pd.pivot_table(df17, values=['bene_count'], index=['npi'])

# step 2 - calculate the average (mean) beneficiaries of all providers
avg_benef_per_prov_b = benef_per_prov_df17_b.mean()

avg_benef_per_prov_b

bene_count    158.349459
dtype: float64

### Problem 2:

#### 2.1 length of average prescription from the total_day_supply for each provider

In [11]:
avg_day_supply_pres_for_each_prov = df17.groupby(df17['npi'])['total_day_supply'].mean()

avg_day_supply_pres_for_each_prov

npi
1003000126    14788
1003000142    58605
1003000167      430
1003000175      150
1003000282     2681
              ...  
1992999825     5566
1992999833     1866
1992999866      288
1992999874     1879
1992999882     7945
Name: total_day_supply, Length: 1162898, dtype: int64

#### 2.2 length of average prescription from the total_claim_count for each provider

In [12]:

avg_claim_count_pres_for_each_prov = df17.groupby(df17['npi'])['total_claim_count'].mean()

avg_claim_count_pres_for_each_prov

npi
1003000126     677
1003000142    1946
1003000167      55
1003000175      18
1003000282      90
              ... 
1992999825     270
1992999833      59
1992999866      15
1992999874      81
1992999882     256
Name: total_claim_count, Length: 1162898, dtype: int64

#### 2.3 median (in days) of the distribution of total_day_supply across all providers

In [13]:
med_day_supply_all_prov = df17['total_day_supply'].median()

med_day_supply_all_prov

4685.0

### Problem 3:

#### 3.1 (a) ratio of beneficiaries with opioid prescriptions to 
#### beneficiaries with antibiotics prescriptions in each state (including rows with null values)

In [14]:
def ratio_op_ab(fn, ln):
    return fn / ln

df17['ratio_bene_opd_antb'] = ratio_op_ab(df17['opioid_bene_count'], df17['antibiotic_bene_count'])
avg_ratio_bene_opd_antb_state = df17.groupby(df17['nppes_provider_state'])['ratio_bene_opd_antb'].mean()

avg_ratio_bene_opd_antb_state

nppes_provider_state
AA    0.463247
AE    0.724034
AK         inf
AL         inf
AP         inf
        ...   
WI         inf
WV         inf
WY         inf
XX    0.662638
ZZ         inf
Name: ratio_bene_opd_antb, Length: 61, dtype: float64

#### 3.1 (b) ratio of beneficiaries with opioid prescriptions to 
#### beneficiaries with antibiotics prescriptions in each state (excluding rows with null values)

In [15]:
df17a = df17[['npi','nppes_provider_state','opioid_bene_count','antibiotic_bene_count']]
df17a.head()

npi nppes_provider_state  opioid_bene_count  antibiotic_bene_count
0  1003000126                   MD               24.0                  104.0
1  1003000142                   OH              200.0                    0.0
2  1003000167                   NV               11.0                   22.0
3  1003000175                   CA                0.0                    NaN
4  1003000282                   TN                0.0                    0.0

In [16]:
df17a.shape

(1162898, 4)

In [17]:
df17b1 = df17a[pd.notnull(df17a['opioid_bene_count'])]

In [18]:
df17b1.shape

(775464, 4)

In [19]:
df17b1.isnull().sum()

npi                           0
nppes_provider_state          0
opioid_bene_count             0
antibiotic_bene_count    234185
dtype: int64

In [20]:
df17a['opioid_bene_count'].isnull().sum()

387434

In [21]:
df17b2 = df17b1[pd.notnull(df17b1['antibiotic_bene_count'])]

In [22]:
df17b2.shape

(541279, 4)

In [23]:
df17b2.isnull().sum()

npi                      0
nppes_provider_state     0
opioid_bene_count        0
antibiotic_bene_count    0
dtype: int64

In [24]:
df17b2.head()

npi nppes_provider_state  opioid_bene_count  antibiotic_bene_count
0  1003000126                   MD               24.0                  104.0
1  1003000142                   OH              200.0                    0.0
2  1003000167                   NV               11.0                   22.0
4  1003000282                   TN                0.0                    0.0
5  1003000407                   PA               11.0                   55.0

In [25]:
df17b2.dtypes

npi                        int64
nppes_provider_state      object
opioid_bene_count        float64
antibiotic_bene_count    float64
dtype: object

In [26]:
df17b3 = df17b2.loc[(df17b2['opioid_bene_count'] != 0.0) & (df17b2['antibiotic_bene_count'] != 0.0)] 

In [27]:
df17b3.shape

(346803, 4)

In [28]:
df17b3['opioid_bene_count'].unique

<bound method Series.unique of 0          24.0
2          11.0
5          11.0
8          51.0
9          36.0
           ... 
1162880    57.0
1162881    95.0
1162885    61.0
1162887    70.0
1162893    27.0
Name: opioid_bene_count, Length: 346803, dtype: float64>

In [29]:
df17b3['antibiotic_bene_count'].unique

<bound method Series.unique of 0          104.0
2           22.0
5           55.0
8           98.0
9           89.0
           ...  
1162880     22.0
1162881    291.0
1162885     63.0
1162887    145.0
1162893     59.0
Name: antibiotic_bene_count, Length: 346803, dtype: float64>

In [30]:
df17b3.head(10)

npi nppes_provider_state  opioid_bene_count  antibiotic_bene_count
0   1003000126                   MD               24.0                  104.0
2   1003000167                   NV               11.0                   22.0
5   1003000407                   PA               11.0                   55.0
8   1003000522                   FL               51.0                   98.0
9   1003000530                   PA               36.0                   89.0
10  1003000597                   OK               77.0                   96.0
16  1003000902                   KY               53.0                  112.0
21  1003001256                   CO               19.0                   14.0
26  1003001587                   CT               23.0                   28.0
31  1003001884                   MI               26.0                   58.0

In [31]:
def ratio_opid_antib(fn, ln):
    return fn / ln

df17b3['ratio_bene_opid_antib'] = ratio_opid_antib(df17b3['opioid_bene_count'], df17b3['antibiotic_bene_count'])


/Applications/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


In [32]:
avg_ratio_bene_opid_antib_state = df17b3.groupby(df17b3['nppes_provider_state'])['ratio_bene_opid_antib'].mean()

avg_ratio_bene_opid_antib_state

nppes_provider_state
AA    0.648546
AE    0.822766
AK    0.979123
AL    1.414440
AP    0.759456
        ...   
WI    0.982913
WV    1.042236
WY    0.916116
XX    1.104397
ZZ    0.761130
Name: ratio_bene_opid_antib, Length: 61, dtype: float64

#### 3.2 assuming each beneficiary attends only a single provider, difference between the largest and smallest ratios

In [34]:
min_ratio = df17b3['ratio_bene_opid_antib'].min()
max_ratio = df17b3['ratio_bene_opid_antib'].max()

diff_bet_min_max = max_ratio - min_ratio
diff_bet_min_max

189.53282828282826

In [35]:
min_ratio

0.012626262626262626

In [36]:
max_ratio

189.54545454545453

### Problem 4:

In [37]:
df_op = df17[['npi','nppes_provider_state','specialty_description','opioid_claim_count','opioid_day_supply']]

In [38]:
df_op.shape

(1162898, 5)

In [39]:
df_op.head()

npi nppes_provider_state specialty_description  opioid_claim_count  \
0  1003000126                   MD     Internal Medicine                25.0   
1  1003000142                   OH        Anesthesiology              1040.0   
2  1003000167                   NV               Dentist                14.0   
3  1003000175                   CA               Dentist                 0.0   
4  1003000282                   TN    Nurse Practitioner                 0.0   

   opioid_day_supply  
0              134.0  
1            29238.0  
2               35.0  
3                0.0  
4                0.0

In [40]:
# 4.1 - states that have surprisingly high supply of opioids, conditioned on specialty 

In [41]:
df_op_supply = df_op[['npi','nppes_provider_state','specialty_description','opioid_day_supply']]
df_op_supply

npi nppes_provider_state specialty_description  \
0        1003000126                   MD     Internal Medicine   
1        1003000142                   OH        Anesthesiology   
2        1003000167                   NV               Dentist   
3        1003000175                   CA               Dentist   
4        1003000282                   TN    Nurse Practitioner   
...             ...                  ...                   ...   
1162893  1992999825                   WA        Otolaryngology   
1162894  1992999833                   NY         Psychoanalyst   
1162895  1992999866                   FL               Dentist   
1162896  1992999874                   VA     Internal Medicine   
1162897  1992999882                   PA   Physician Assistant   

         opioid_day_supply  
0                    134.0  
1                  29238.0  
2                     35.0  
3                      0.0  
4                      0.0  
...                    ...  
1162893              131.0  
1162894                0.0  
1162895                NaN  
1162896                NaN  
1162897                0.0  

[1162898 rows x 4 columns]

In [42]:
df_op_supply1 = df_op_supply[pd.notnull(df_op_supply['opioid_day_supply'])]

In [43]:
df_op_supply1.head()

npi nppes_provider_state specialty_description  opioid_day_supply
0  1003000126                   MD     Internal Medicine              134.0
1  1003000142                   OH        Anesthesiology            29238.0
2  1003000167                   NV               Dentist               35.0
3  1003000175                   CA               Dentist                0.0
4  1003000282                   TN    Nurse Practitioner                0.0

In [44]:
df_op_supply2 = df_op_supply1.loc[(df_op_supply1['opioid_day_supply'] != 0.0)]

In [45]:
df_op_supply2.shape

(491831, 4)

In [46]:
df_op_supply2['opioid_day_supply'].unique

<bound method Series.unique of 0            134.0
1          29238.0
2             35.0
5            622.0
8           3509.0
            ...   
1162885      435.0
1162887     5227.0
1162891     1952.0
1162892       85.0
1162893      131.0
Name: opioid_day_supply, Length: 491831, dtype: float64>

In [47]:
df_op_supply2['nppes_provider_state'].unique()

array(['MD', 'OH', 'NV', 'PA', 'FL', 'OK', 'KY', 'CO', 'CA', 'TX', 'CT',
       'MI', 'NJ', 'MA', 'MS', 'NY', 'SC', 'MN', 'WA', 'DC', 'NC', 'IA',
       'AZ', 'LA', 'VA', 'AR', 'HI', 'OR', 'KS', 'RI', 'WI', 'ID', 'IN',
       'NM', 'TN', 'PR', 'GA', 'WV', 'VT', 'AL', 'ME', 'ND', 'AE', 'MO',
       'IL', 'UT', 'NH', 'NE', 'DE', 'MT', 'WY', 'SD', 'AK', 'ZZ', 'AP',
       'MP', 'AA', 'GU', 'VI', 'XX', 'AS'], dtype=object)

In [49]:
df_op_sort_by_state_specialty = df_op_supply2.sort_values(by=['nppes_provider_state',
                                                              'specialty_description',
                                                              'opioid_day_supply'], ascending = False)

In [50]:
df_op_sort_by_state_specialty.iloc[0:60,0:5]

npi nppes_provider_state  \
314445   1265986087                   ZZ   
80950    1063934396                   ZZ   
792103   1689029159                   ZZ   
996041   1851594634                   ZZ   
22357    1013975481                   ZZ   
795317   1689644619                   ZZ   
350926   1306071816                   ZZ   
65437    1053571794                   ZZ   
238970   1205819497                   ZZ   
912280   1780679258                   ZZ   
938502   1801875836                   ZZ   
796029   1689661845                   ZZ   
1157972  1992780407                   ZZ   
1054977  1902888563                   ZZ   
350381   1306043484                   ZZ   
967251   1831164607                   ZZ   
731942   1629468145                   ZZ   
184882   1154774933                   ZZ   
746982   1649254392                   ZZ   
642710   1558358077                   ZZ   
725082   1629097506                   ZZ   
892147   1760726707                   ZZ   
618119   1538128665                   ZZ   
278211   1235576695                   ZZ   
998726   1851757348                   ZZ   
12392    1013025915                   ZZ   
1145528  1982699302                   ZZ   
665686   1578542957                   ZZ   
1133466  1972661726                   ZZ   
956078   1821076738                   ZZ   
357285   1306854732                   ZZ   
130529   1114140811                   ZZ   
623128   1538289293                   ZZ   
224553   1194701524                   ZZ   
684087   1588833677                   ZZ   
1062551  1912241035                   ZZ   
147402   1124318555                   ZZ   
770946   1669467213                   ZZ   
1088796  1932389574                   ZZ   
972682   1831384247                   ZZ   
25020    1023043734                   ZZ   
781525   1679532550                   ZZ   
808675   1699752410                   ZZ   
292140   1255328068                   ZZ   
1005652  1861582801                   ZZ   
25823    1023067295                   ZZ   
618600   1538140124                   ZZ   
148956   1124415773                   XX   
358905   1306903265                   XX   
292165   1255328753                   XX   
358634   1306894159                   XX   
555158   1477785731                   XX   
983259   1841422094                   WY   
318511   1275592024                   WY   
969098   1831209253                   WY   
347663   1295896751                   WY   
282871   1245293455                   WY   
351974   1306139050                   WY   
864255   1740291137                   WY   
1125105  1962765586                   WY   

                                     specialty_description  opioid_day_supply  
314445                                             Urology              146.0  
80950    Student in an Organized Health Care Education/...               63.0  
792103   Student in an Organized Health Care Education/...               54.0  
996041                                 Physician Assistant            13028.0  
22357                                  Physician Assistant             1378.0  
795317                                 Physician Assistant              391.0  
350926                                 Physician Assistant               62.0  
65437                                       Otolaryngology              224.0  
238970                                      Otolaryngology              118.0  
912280                                  Orthopedic Surgery              134.0  
938502                         Oral Surgery (Dentist only)             1309.0  
796029                         Oral Surgery (Dentist only)              261.0  
1157972                            Obstetrics & Gynecology               79.0  
1054977                                  Internal Medicine             6652.0  
350381                                   Internal Medicine             3117.0  
967251                                   Internal M

In [52]:
# 4.2 - average length of an opioid prescription for each provider
df_op_prov_pres = pd.pivot_table(df_op, values=['opioid_day_supply'], index=['npi'])
df_op_prov_pres

opioid_day_supply
npi                          
1003000126              134.0
1003000142            29238.0
1003000167               35.0
1003000175                0.0
1003000282                0.0
...                       ...
1992999791             1952.0
1992999817               85.0
1992999825              131.0
1992999833                0.0
1992999882                0.0

[834341 rows x 1 columns]

In [53]:
df_op_prov_pres_sort = df_op_prov_pres.sort_values(by=['npi'], ascending = True)
df_op_prov_pres_sort

opioid_day_supply
npi                          
1003000126              134.0
1003000142            29238.0
1003000167               35.0
1003000175                0.0
1003000282                0.0
...                       ...
1992999791             1952.0
1992999817               85.0
1992999825              131.0
1992999833                0.0
1992999882                0.0

[834341 rows x 1 columns]

In [ ]:
# 4.3 - for each (state, specialty) pair with at least 100 providers,
#       calculate the average of this value across all providers

In [ ]:
# 4.4 - ratio of this value to an equivalent quantity calculated from providers in each specialty across all states


In [ ]:
# 4.5 - What is the largest such ratio?

### Problem 5:

In [54]:
op_agelis = df17[['npi',
'total_claim_count',
'total_claim_count_ge65',
'lis_claim_count']]

In [55]:
op_agelis.head()

npi  total_claim_count  total_claim_count_ge65  lis_claim_count
0  1003000126                677                   516.0            305.0
1  1003000142               1946                   881.0           1184.0
2  1003000167                 55                     NaN              NaN
3  1003000175                 18                     NaN              NaN
4  1003000282                 90                    65.0             73.0

In [57]:
op_agelis1 = op_agelis[pd.notnull(op_agelis['total_claim_count'])]
op_agelis1.head()

npi  total_claim_count  total_claim_count_ge65  lis_claim_count
0  1003000126                677                   516.0            305.0
1  1003000142               1946                   881.0           1184.0
2  1003000167                 55                     NaN              NaN
3  1003000175                 18                     NaN              NaN
4  1003000282                 90                    65.0             73.0

In [58]:
op_agelis1.shape

(1162898, 4)

#### 5.1 For each provider where the relevant columns are not suppressed, 
####       work out the fraction of claims for beneficiaries age 65 and older.

In [59]:
op_agelis2 = op_agelis1[pd.notnull(op_agelis1['total_claim_count_ge65'])]
op_agelis2.head()

npi  total_claim_count  total_claim_count_ge65  lis_claim_count
0  1003000126                677                   516.0            305.0
1  1003000142               1946                   881.0           1184.0
4  1003000282                 90                    65.0             73.0
5  1003000407               2788                  2700.0           1990.0
6  1003000423                200                   118.0             70.0

In [60]:
op_agelis2.shape

(910442, 4)

In [61]:
op_agelis2a = op_agelis2.loc[
    (op_agelis2['total_claim_count'] != 0.0) & 
    (op_agelis2['total_claim_count_ge65'] != 0.0)]

In [62]:
op_agelis2a.shape

(901283, 4)

In [63]:
op_agelis2a['frac_total_claim_count_ge65'] = op_agelis2a['total_claim_count_ge65'] / op_agelis2a['total_claim_count']

/Applications/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [64]:
op_agelis2a.head()

npi  total_claim_count  total_claim_count_ge65  lis_claim_count  \
0  1003000126                677                   516.0            305.0   
1  1003000142               1946                   881.0           1184.0   
4  1003000282                 90                    65.0             73.0   
5  1003000407               2788                  2700.0           1990.0   
6  1003000423                200                   118.0             70.0   

   frac_total_claim_count_ge65  
0                     0.762186  
1                     0.452724  
4                     0.722222  
5                     0.968436  
6                     0.590000

In [65]:
op_agelis2a_frac_by_prov = op_agelis2a.groupby(op_agelis2a['npi'])['frac_total_claim_count_ge65'].mean()

In [66]:
op_agelis2a_frac_by_prov

npi
1003000126    0.762186
1003000142    0.452724
1003000282    0.722222
1003000407    0.968436
1003000423    0.590000
                ...   
1992999791    0.525140
1992999825    0.848148
1992999833    1.000000
1992999874    0.814815
1992999882    0.210938
Name: frac_total_claim_count_ge65, Length: 901283, dtype: float64

#### 5.2 For each provider where the relevant columns are not suppressed, 
####       the fraction of claims for beneficiaries with a low-income subsidy.

In [67]:
op_agelis3 = op_agelis1[pd.notnull(op_agelis1['lis_claim_count'])]
op_agelis3.head()

npi  total_claim_count  total_claim_count_ge65  lis_claim_count
0  1003000126                677                   516.0            305.0
1  1003000142               1946                   881.0           1184.0
4  1003000282                 90                    65.0             73.0
5  1003000407               2788                  2700.0           1990.0
6  1003000423                200                   118.0             70.0

In [68]:
op_agelis3.shape

(922025, 4)

In [69]:
op_agelis3a = op_agelis3.loc[
    (op_agelis3['lis_claim_count'] != 0.0) & 
    (op_agelis3['total_claim_count'] != 0.0)]

In [70]:
op_agelis3a.shape

(858413, 4)

In [71]:
op_agelis3a['frac_lis_claim_count'] = op_agelis3a['lis_claim_count'] / op_agelis3a['total_claim_count']

/Applications/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [72]:
op_agelis3a.head()

npi  total_claim_count  total_claim_count_ge65  lis_claim_count  \
0  1003000126                677                   516.0            305.0   
1  1003000142               1946                   881.0           1184.0   
4  1003000282                 90                    65.0             73.0   
5  1003000407               2788                  2700.0           1990.0   
6  1003000423                200                   118.0             70.0   

   frac_lis_claim_count  
0              0.450517  
1              0.608428  
4              0.811111  
5              0.713773  
6              0.350000

In [73]:
op_agelis3a_frac_by_prov = op_agelis3a.groupby(op_agelis3a['npi'])['frac_lis_claim_count'].mean()

In [74]:
op_agelis3a_frac_by_prov

npi
1003000126    0.450517
1003000142    0.608428
1003000282    0.811111
1003000407    0.713773
1003000423    0.350000
                ...   
1992999759    0.725049
1992999791    0.810056
1992999825    0.214815
1992999874    0.345679
1992999882    0.859375
Name: frac_lis_claim_count, Length: 858413, dtype: float64

#### 5.3 (a) What is the Pearson correlation coefficient between these values - 
####            the fraction of claims for beneficiaries age 65 and older and
####            the fraction of claims for beneficiaries with a low-income subsidy?

In [75]:
op_agelis4 = op_agelis1[pd.notnull(op_agelis1['total_claim_count_ge65'])]
op_agelis4.head()

npi  total_claim_count  total_claim_count_ge65  lis_claim_count
0  1003000126                677                   516.0            305.0
1  1003000142               1946                   881.0           1184.0
4  1003000282                 90                    65.0             73.0
5  1003000407               2788                  2700.0           1990.0
6  1003000423                200                   118.0             70.0

In [76]:
op_agelis4.shape

(910442, 4)

In [77]:
op_agelis4a = op_agelis4[pd.notnull(op_agelis4['lis_claim_count'])]
op_agelis4a.head()

npi  total_claim_count  total_claim_count_ge65  lis_claim_count
0  1003000126                677                   516.0            305.0
1  1003000142               1946                   881.0           1184.0
4  1003000282                 90                    65.0             73.0
5  1003000407               2788                  2700.0           1990.0
6  1003000423                200                   118.0             70.0

In [78]:
op_agelis4a.shape

(834684, 4)

In [79]:
op_agelis4b = op_agelis4a.loc[
    (op_agelis4a['lis_claim_count'] != 0.0) & 
    (op_agelis4a['total_claim_count'] != 0.0)]

In [80]:
op_agelis4b.shape

(788529, 4)

In [81]:
op_agelis4c = op_agelis4b.loc[(op_agelis4b['total_claim_count_ge65'] != 0.0)]

In [82]:
op_agelis4c.shape

(781327, 4)

In [83]:
op_agelis4c.head()

npi  total_claim_count  total_claim_count_ge65  lis_claim_count
0  1003000126                677                   516.0            305.0
1  1003000142               1946                   881.0           1184.0
4  1003000282                 90                    65.0             73.0
5  1003000407               2788                  2700.0           1990.0
6  1003000423                200                   118.0             70.0

#### 5.3 (b) What is the Pearson correlation coefficient between these values - 
####            the fraction of claims for beneficiaries age 65 and older and
####            the fraction of claims for beneficiaries with a low-income subsidy?

In [84]:
op_agelis4c['frac_total_claim_count_ge65'] = op_agelis4c['total_claim_count_ge65'] / op_agelis4c['total_claim_count']

/Applications/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [85]:
op_agelis4c.head()

npi  total_claim_count  total_claim_count_ge65  lis_claim_count  \
0  1003000126                677                   516.0            305.0   
1  1003000142               1946                   881.0           1184.0   
4  1003000282                 90                    65.0             73.0   
5  1003000407               2788                  2700.0           1990.0   
6  1003000423                200                   118.0             70.0   

   frac_total_claim_count_ge65  
0                     0.762186  
1                     0.452724  
4                     0.722222  
5                     0.968436  
6                     0.590000

In [86]:
op_agelis4c['frac_lis_claim_count'] = op_agelis4c['lis_claim_count'] / op_agelis4c['total_claim_count']

/Applications/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [87]:
op_agelis4c.head()

npi  total_claim_count  total_claim_count_ge65  lis_claim_count  \
0  1003000126                677                   516.0            305.0   
1  1003000142               1946                   881.0           1184.0   
4  1003000282                 90                    65.0             73.0   
5  1003000407               2788                  2700.0           1990.0   
6  1003000423                200                   118.0             70.0   

   frac_total_claim_count_ge65  frac_lis_claim_count  
0                     0.762186              0.450517  
1                     0.452724              0.608428  
4                     0.722222              0.811111  
5                     0.968436              0.713773  
6                     0.590000              0.350000

In [88]:
op_agelis4e = op_agelis4c.drop('npi', axis = 1)

In [89]:
op_agelis4e.head()

total_claim_count  total_claim_count_ge65  lis_claim_count  \
0                677                   516.0            305.0   
1               1946                   881.0           1184.0   
4                 90                    65.0             73.0   
5               2788                  2700.0           1990.0   
6                200                   118.0             70.0   

   frac_total_claim_count_ge65  frac_lis_claim_count  
0                     0.762186              0.450517  
1                     0.452724              0.608428  
4                     0.722222              0.811111  
5                     0.968436              0.713773  
6                     0.590000              0.350000

In [90]:
pearsoncorr = op_agelis4e.corr(method='pearson')
pearsoncorr

total_claim_count  total_claim_count_ge65  \
total_claim_count                     1.000000                0.976948   
total_claim_count_ge65                0.976948                1.000000   
lis_claim_count                       0.855426                0.805184   
frac_total_claim_count_ge65           0.141495                0.242831   
frac_lis_claim_count                  0.002223               -0.059901   

                             lis_claim_count  frac_total_claim_count_ge65  \
total_claim_count                   0.855426                     0.141495   
total_claim_count_ge65              0.805184                     0.242831   
lis_claim_count                     1.000000                    -0.008114   
frac_total_claim_count_ge65        -0.008114                     1.000000   
frac_lis_claim_count                0.245342                    -0.583849   

                             frac_lis_claim_count  
total_claim_count                        0.002223  
total_claim_count_ge65                  -0.059901  
lis_claim_count                          0.245342  
frac_total_claim_count_ge65             -0.583849  
frac_lis_claim_count                     1.000000

##### pearson correlation between 'total_claim_count_ge65' & 'lis_claim_count' = 0.8052

##### pearson correlation between 'frac_total_claim_count_ge65' & 'frac_lis_claim_count' = -0.5838

In [ ]:
# Problem 6:
# 6.1 - For each provider for whom the information is not suppressed, 
# figure out the average cost per day of prescriptions in both 2016 and 2017. 
# 6.2 - Use this to estimate the inflation rate for daily prescription costs per provider. 
# 6.3 - What is the average inflation rate across all providers?

In [ ]:
# Problem 7:
# 7.1 - Consider all providers with a defined specialty in both years. Find the fraction of providers who left each specialty between 2016 and 2017. 
# 7.2 - What is the largest such fraction, when considering specialties with at least 1000 proviers in 2016? 
# Note that some specialties have a fraction of 1 due to specialty name changes between 2016 and 2017; 
# disregard these specialties in calculating your answer.